### ISSUE TITLE
Re-enable the last few eslint rules

### ISSUE THREAD

**Issue by bovine3dom** [tridactyl/tridactyl#2513]
We have a few sensible eslint rules turned off. Fixing these probably isn't too difficult but might be a little time-consuming.

Recommended workflow: 

```
# In one terminal
ls .eslintrc.js src/**.js **.ts | entr -s '$(yarn bin)/eslint --fix src/'
# In another: make your fixes and re-enable rules
```

- [ ] https://github.com/tridactyl/tridactyl/blob/3a80fd03cbd84111958ee200cb9fd91c171f40b7/.eslintrc.js#L85

- [ ] https://github.com/tridactyl/tridactyl/blob/3a80fd03cbd84111958ee200cb9fd91c171f40b7/.eslintrc.js#L108

- [ ] https://github.com/tridactyl/tridactyl/blob/3a80fd03cbd84111958ee200cb9fd91c171f40b7/.eslintrc.js#L124

- [ ] https://github.com/tridactyl/tridactyl/blob/3a80fd03cbd84111958ee200cb9fd91c171f40b7/.eslintrc.js#L125

- [ ] https://github.com/tridactyl/tridactyl/blob/3a80fd03cbd84111958ee200cb9fd91c171f40b7/.eslintrc.js#L126

- [ ] https://github.com/tridactyl/tridactyl/blob/3a80fd03cbd84111958ee200cb9fd91c171f40b7/.eslintrc.js#L127

- [ ] https://github.com/tridactyl/tridactyl/blob/3a80fd03cbd84111958ee200cb9fd91c171f40b7/.eslintrc.js#L157

**Comment by MyGitUsername** [comment]
I can start chipping away at this issue, one rule at a time.

**Comment by bovine3dom** [comment]
That would be great : )

**Comment by MyGitUsername** [comment]
If anyone else is interested in contributing to this issue, please feel welcome :)  
I haven't made any real progress here yet.  I'm planning to enable just one rule at a time and commit in small chunks to keep the work manageable as advised.

**Comment by Rummskartoffel** [comment]
I think `@typescript-eslint/explicit-function-return-type` might also be worth considering adding to the backlog.

**Comment by bovine3dom** [comment]
What's wrong with letting typescript work it out for us?

If there is an option to throw an error if it can't work it out and falls back to `any`, I'd much prefer that.

**Comment by Rummskartoffel** [comment]
There is `noImplicitAny`, but I think that only applies to unannotated variable declarations like `let foo` as opposed to `let foo = "bar"` or `let foo: string`, while functions always infer their type based on what they do or don't return. If the type of one of the possible return values is a subset of the other value's type, then that more general type is the type of the function otherwise it's a union type of all the possible return types.
After trying out that linter rule on hinting.ts, though, I agree with you; it would be overkill and lead to having to add a bunch of unnecessary type annotations, which would make for a lot of work and result in a lot of clutter. What I'd like would be a linter rule or compiler option that only disallows inferred union return types and maybe even inferred generalisations – because those are probably not intended and should be marked, IMO, if they are –, but such a thing doesn't seem to exist.

**Comment by bovine3dom** [comment]
We should change the no-unused-vars rule to the experimental one. The current one doesn't support imported types.

```diff
diff --git a/.eslintrc.js b/.eslintrc.js
index aa8d3a7c..1c6d0397 100644
--- a/.eslintrc.js
+++ b/.eslintrc.js
@@ -131,12 +131,13 @@ module.exports = {
                 "allowTernary": true,
             }
         ],
-        "@typescript-eslint/no-unused-vars": [
+        "@typescript-eslint/no-unused-vars-experimental": [
             "error",
             {
-                "args": "none",
-            }
+                "ignoreArgsIfArgsAfterAreUsed": true,
+            },
         ],
+        "@typescript-eslint/no-unused-vars": "off",
         "@typescript-eslint/no-use-before-define": "off",
         "@typescript-eslint/no-var-requires": "error",
         "@typescript-eslint/prefer-as-const": "error",
```

Unfortunately, the new rule doesn't allow for named unused variables, which I quite like for documentation purposes. There's a reasonable amount of stuff to fix.

**Comment by callmetushar123** [comment]
@bovine3dom Hey I need to change those mentioned rules to on right?

**Comment by bovine3dom** [comment]
@callmetushar123 yeah, the ones mentioned in the original post and then fix the errors. If there are any other ones you know about that you think might be beneficial, feel free to enable them too :)

I suggest you just do one at once because you will find that lots of them need lots of changes...
